### ISSUE TITLE
bug: concept self-references

### ISSUE THREAD

**Issue by AlanSimmons** [x-atlas-consortia/ubkg-etl#9]
@benstear 
# Statement of problem

The OWLNETS-UMLS-GRAPH script introduces self-references to concepts, in which a concept has a reference to itself. This results in "rosettes" in the knowledge graph in response to queries such as 
```match (p1:Concept)-[]-(p2:Concept) where p1.CUI=p2.CUI return *```--e.g.,
<img width="697" alt="image" src="https://user-images.githubusercontent.com/10928372/227943565-99273100-e4a0-4a80-9067-3cf936483a6b.png">

Cursory examination of the self-referential relationships indicate that they have SABs from ingested ontologies--i.e., they were created by the script.

According to @computationdoc and existing in-line documentation, the script had accounted for self-references in two places:

1. Immediately after reading in the edges file
2. When assigning a preferred CUI for each ingested node from the nodes file.

The logic in the second part of the script involved:

- building for each node a list of CUIs (cuis) ordered by preference

1. nodeCUIs - the UMLS CUI for a cross-reference that existed previously in the ontology CSVs
2. CUI_CODEs - the UMLS CUI for the node that existed previously in the ontology CSVs
3. XrefCUIs - the non-UMLS CUIs for a cross-reference that existed previously in the ontology CSVs
4. base64cui - the base64 CUI (base64 encoding of the node id) for a node that does not currently exist in the ontology CSVs

Looping through the list as follows:
```
node_idCUIs = []
nmCUI = []
for index, rows in node_metadata.iterrows():
    addedone = False
    for x in rows.cuis:
        if ((x in node_idCUIs) | (addedone == True)):
            dummy = 0
        else:
            nmCUI.append(x)
            node_idCUIs.append(x)
            addedone = True
    # JAS 17 oct 2022
    # For the edge case in which multiple nodes have the same cross-reference AND the
    # cross-reference maps to a single CUI (the case for around 20 nodes in MP--e.g., all
    # those that share CL 0000959 as a cross-reference), only one of the nodes will be
    # assigned the CUI of the cross-reference. The rest of the nodes will lose the
    # cross-reference.
    # The following block keeps the nmCUI list the same size as node_metadata['CUI'].
    if addedone == False:
        nmCUI.append('')
        node_idCUIs.append('')

node_metadata['CUI'] = nmCUI
```

I added the code at the bottom of the block to account for errors that occurred when ingesting MP.

Because this loop can consume a prohibitively amount of resources for large sets of nodes, I attempted to replace the loop with a simpler call:

```
# Assign the CUI in order of precedence. This should be the first element in the list of cuis.
#node_metadata['CUI']=node_metadata['cuis'].str[0]
```

During discussion of this change, we discovered that self-references had been re-introduced. However, when I reverted to the original loop, the self-references were still in the graph. This means that either the original code did not address the issue of self-references or I re-introduced self-references as part of other changes to the script. 

# Possible solution options to explore
I still think that the simpler code is equivalent to the original loop because the calculated cuis field orders potential CUIs such that the first element in a node's list is the preferred CUI. However, I will verify this by comparing the intermediate DataFrames generated by each code.

I also think that it may be possible to eliminate self-references at the point of appending to CUI-CUIs.CSV.

**Comment by AlanSimmons** [comment]
# Code to remove self-references

I added a line of code that removed the self-referenced edges before writing to CUI-CUIs.csv:

_start of script_
...
_code to assign preferred CUI to node_metadata_
_code that merges edgelist to node_metadata_

```
# -------------------------------------------------
# JAS 27 MAR 2023
# Remove "self-references"--i.e., assertions in which the CUIs for subject and object nodes are identical.

print('Removing self-reference edges...')
edgelist = edgelist[edgelist['CUI1']!=edgelist['CUI2']]

```
_code that appends data from edgelist to CUI-CUIs.csv_
...
_end of script_

I compared CUI-CUIs.csv before and after for the new code for the ingestion of UBERON. Self-refererences (where both of the CUIs were identical) were removed in the "after" version.

**Comment by AlanSimmons** [comment]
# Analysis: Comparison of the CUI assignment algorithms

The script assigns a preferred CUI for each node in the node list.

Prior to the selection, the script derives a list of CUIs in order of preference:

1. nodeCUIs - the UMLS CUI for a cross-reference that existed previously in the ontology CSVs
2. CUI_CODEs - the UMLS CUIs for the node that existed previously in the ontology CSVs
3. XrefCUIs - the non-UMLS CUIs for a cross-reference that existed previously in the ontology CSVs
4. base64cui - the base64 CUI (base64 encoding of the node id) for a node that does not currently exist in the ontology CSVs

CUI_CODEs and XrefCUIs can be lists.

The important factor here is that more than one node can link to a CUI, especially as cross-references. 

# Original algorithm
The original algorithm matches each node to a unique CUI by keeping track of assignments:

```
# iterate to select one CUI from cuis in row order - we ensure each node_id has its own distinct CUI
# each node_id is assigned one CUI distinct from all others' CUIs to ensure no self-reference in edgelist

node_idCUIs = []
nmCUI = []
itest = 0
for index, rows in node_metadata.iterrows():
    addedone = False
    for x in rows.cuis:
        if ((x in node_idCUIs) | (addedone == True)):
            dummy = 0
        else:
            nmCUI.append(x)
            node_idCUIs.append(x)
            addedone = True
    if addedone == False:
        print('node:',rows.node_id,'has no available CUIs; assigning blank')
        nmCUI.append('')
        node_idCUIs.append('')

node_metadata['CUI'] = nmCUI
```

# New algorithm
The new algorithm simply assigns the first CUI in the ordered list.
```
node_metadata['CUI']=node_metadata['cuis'].str[0]
```

# Example of different results
When ingesting UBERON, the script calculates the following ordered list of CUIs:
['C0162412', 'C0024955', 'VUJFUk9OIDAwMDAzNzc=']
The list has two cross-referenced CUIs and the base64 CUI.

The first algorithm selects C0024955 because a node earlier in the list was assigned C0162412.
The second algorithm selects C0162412

# First algorithm pros and cons
## Pro: every node is assigned a unique CUI, including cross-references
## Cons:
1. The algorithm is slow.
2. The algorithm requires considerable resources, especially for large sets of nodes. Ben has proposed an optimization that replaces the list with a set.
3. It is possible that a node will not have a CUI available. Cases were first identified in the ingestion of MP. 

There are at least two edge cases in which a node will not be assigned a CUI:
1. Node 1 defines a cross-reference to Node 2 before Node2 defines its CUI AND Node2 only has a base64 CUI. Node2's CUI is used for Node1 and Node2 does not have a CUI
2. Node3 is defined with only a base64 CUI and Node 1 and Node 2 define a cross-reference to Node 3. Node 3 gets its base64 CUI and the two other nodes do not have a CUI available.

MP currently has around 20 CL nodes that are not assigned CUIs.

Originally, ingesting MP would cause the script to fail, because the node_metadata and nmCUI lists would be of different sizes. I added the code that assigned a blank CUI to nmCUI; however, this results in lines like this in CUI-CODEs.csv:
:START_ID,:END_ID
,CL 0002357

AZ presents another challenge: multiple nodes that have the same cross-reference. In this case, a node can lose its cross-reference.

For example, AZ:0000004, AZ:0000025, and AZ:0000080 share a cross-reference with CL:0000236, and CL:0000236 cross-references to 2 CUIs. The script calculates the following CUI lists and assignments:
AZ code | CUI list | CUI
---|---|---
AZ:0000004 |['C1267837', 'C0004561', 'QVogMDAwMDAwNA=='] |C1267837
AZ:0000025|['C0004561', 'C1267837', 'QVogMDAwMDAyNQ==']|C0004561
AZ:0000080|['QVogMDAwMDA4MA==', 'C1267837', 'C0004561']|QVogMDAwMDA4MA==

In other words, the third AZ node loses its cross-reference.

**Comment by AlanSimmons** [comment]
# Verification of new functionality
I regenerated a knowledge graph using the script that had the new changes:

1. Code that removed self-references before writing to CUI-CUIs.csv.
2. The simpler algorithm for preferred CUI assignment.

As the images in the prior post show, the simpler algorithm corrects some issues related to CUI assignment. 

The knowledge graph also no longer contains self-references.

I will refresh the HuBMAP and Data Distillery knowledge graphs with CSVs that are generated with the revisions to the script.

**Comment by AlanSimmons** [comment]
Completed additional regression testing, checking that nodes and edges established as expected.

[Testing tracker](https://docs.google.com/spreadsheets/d/1Z2x5ERAVahaGJII8dhWgv0XmxiWL4nT5Ks89Dv7jpKU/edit?usp=sharing)
